### ISSUE TITLE
Support in pyugrid for node_coordinates other than latitude and longitude

### ISSUE THREAD

**Issue by bhlevca** [pyugrid/pyugrid#145]
Hello,

I am working on a project to bring Python post processing tools to 
[Delft3D-FM ](https://www.deltares.nl/en/software/delft3d-flexible-mesh-suite/), which will be part of the Open Source effort of the [Open Earth Tools](https://publicwiki.deltares.nl/display/OET/OpenEarth). 

I am not affiliated with Deltares, but I am a Python proponent and I am helping them out and integrating it into my preferred tools. 

I am using pyugrid for the map files and so far it is working great except two things:

1. triangular cells only (but I create triangular meshes only at the moment so I am OK, but I hope for support for other shapes in the near future)
2.  pyugrid requires "latitude" and "longitude" as coord_vars, thus grids with UTM coordinates are throwing an exception. I worked around this issue by subclassing pugrid.UGrid and 2 mehtods from red_netcdf.

Perhaps issue 1 is harder to implement, but the second one may be worth while to consider addressing, I am passing along Deltares' request: 
         
> The UGRID conventions document does NOT require the use of latitude longitude, see:
> [UGRID conventions](http://ugrid-conventions.github.io/ugrid-conventions/#ugrid-conventions-v10)
>
> go to : “(latitude, longitude, and optional elevation or other coordinates)”
> Any requirement of lat/lon comes from CF conventions (which UGRID will often be combined with, of course).
> 
> ·As such, the lat/lon checking should not be a required/blocking part of the pyugrid code.
> An idea could be to make the CF-checking in pyugrid configurable optional, or remove it altogether.

 Please let me know what you think.

Thanks
Bogdan

**Comment by ChrisBarker-NOAA** [comment]
Thanks for the note -- and your work!

> triangular cells only (but I create triangular meshes only at the moment so I am OK, but I hope for support for other shapes in the near future)

yes -- that is the idea -- it only takes someone to write the code. We tried to not hard-code triangles in there too much. If we did, it was an error. But someone that needs it is going to need to write the code -- but I'm looking forward to reviewing a PR :-)

> pyugrid requires "latitude" and "longitude" as coord_vars, thus grids with UTM coordinates are throwing an exception. I worked around this issue by subclassing pugrid.UGrid and 2 methods from red_netcdf.

The UGrid class simply uses "nodes". It is probably making assumptions aout floating point values and orthogonal coordinates, but it should be able to handle UTM, etc jsut fine.

So it's the netcdf reading code that needs an update. As long as it allows only CF-compliant coordinates, then I think that would be a great change.

However, IF there are arbitrary coordinate systems, the the UGrid class should keep that information, and appropriately write it out to netCDF as well. Off the op of my head, maybe a node_attributes attribute? or more specifically a crs attribute? I'd have to look at CF to see how/where you specify the CRS -- but we should follow its lead wherever we can.

PRs happily accepted!

Let us know if you need help figuring out where in the code to make changes.

-CHB

**Comment by bhlevca** [comment]
Yes, you are right, `read_netcdf` that causes the problem, specifically in `load_grid_from_nc_dataset`
```
        for var in coord_vars:
           [...]
            if standard_name == 'latitude':
                nodes[:, 1] = var[:]
            elif standard_name == 'longitude':
                nodes[:, 0] = var[:]
            else:
                raise ValueError('Node coordinates standard_name is neither '
                                 '"longitude" nor "latitude" ')
```
This standard name should not be enforced or made configurable.

I am lost in acronyms, what kind of PR are you referring to? 

 Did you mean that I should go ahead and make the change?  I don't know if I can commit to another job, but  I made and make changes in my subclass and derived code and I can post my findings. I create unit tests for new code and I am afraid to promise maintaining extra code as I am spreading too thin. However, I can help with testing and submitting patches and suggestions.  

With the other issue, (triangular -> quad, pentagon, hexagon) I may have a go at it but only in a couple of months as I am on a tight schedule at the moment.

Suggestions for places in the code where I can address my issues are welcome.

Bogdan

**Comment by arthurvd** [comment]
> **@ChrisBarker-NOAA wrote:**
> I'd have to look at CF to see how/where you specify the CRS -- but we should follow its lead wherever we can.

In CF-1.6(/1.7) the specification of the underlying coordinate reference system/grid mapping of the grid (auxiliary) coordinates is optional, but when specified, this should always be with all of the *data* variables defined on those grids.
http://cfconventions.org/cf-conventions/cf-conventions.html#grid-mappings-and-projections :
>"When the coordinate variables for a horizontal grid are not longitude and latitude, it is required that the true latitude and longitude coordinates be supplied via the `coordinates` attribute. If *in addition* it is desired to describe the mapping between the given coordinate variables and the true latitude and longitude coordinates, the attribute `grid_mapping` may be used to supply this description. This attribute is attached to **data** variables[..]"

**Comment by ChrisBarker-NOAA** [comment]
OK, the trick here is that we've got to look for *something* to know that that particular variable is the x and y coordinates -- what does CF use for, for example, UTM coordinates.

a PR is a "Pull Request" -- it how gitHub lets you suggest changes to code to the developers. THere  is a lot of documentation on the web about how to do this -- here's one I found with a quick google:

http://kbroman.org/github_tutorial/pages/fork.html

So yes -- I'm suggesting changing the code :-) -- in your own fork start with. Rule of thumb -- if it adds useful functionality, and doesn't break/change the existing API, and it's clean pythonic code, then we'll merge it.

Posting your extensions is helpful, but it's more helpful if you do the integration with the package too :-) -- but we all only have so much time.

> I create unit tests for new code 

Oh, and "has unit test" -- :-)

>  I am afraid to promise maintaining extra code as I am spreading too thin. 

Making a contribution is not a promise to maintain -- if it's smallish and fits it well with the existing code, we're happy to maintain it from there.

**Comment by ChrisBarker-NOAA** [comment]
"""
In CF-1.6(/1.7) the specification of the underlying coordinate reference system/grid mapping of the grid (auxiliary) coordinates is optional, but when specified, this should always be with all of the data variables defined on those grids.
"""
hmm, a bit ugly, but makes sense.

So, without having thought this out carefully, for UGrid I think it's easier. only the nodes use the CRS -- everything else is mapped to those nodes.

So I think the CRS info should be stored on the UGRid object (optional???, but defaulting to lat-lon WGS84), This is can be read from / written to the netCDF.

However, if I understand this right:

"""
When the coordinate variables for a horizontal grid are not longitude and latitude, it is required that the true latitude and longitude coordinates be supplied via the coordinates attribute.
"""

Then CF DOES require latitude and longitude -- and as we were trying to have UGRID be an extension to CF, then it does too.

Personally, I think relaxing this requirement is fine for pyugrid, if not for the UGRID spec --that's a discussion for another repo...

**Comment by arthurvd** [comment]
> **@ChrisBarker-NOAA wrote:**
> Then CF DOES require latitude and longitude -- and as we were trying to have UGRID be an extension to CF, then it does too.

Fortunately not, UGRID states:
> Note: NetCDF files using this convention can be be given the global attribute Conventions = 'CF-1.6, UGRID-1.0' if they are CF- and UGRID-compliant, **or just Conventions = 'UGRID-1.0' if they are not CF-compliant**. 
(https://github.com/ugrid-conventions/ugrid-conventions)

.
> So, without having thought this out carefully, for UGrid I think it's easier. only the nodes use the CRS -- everything else is mapped to those nodes

Not per se, there's actually also optional edge coordinates and face coordinates, and the data variables defined on those locations (e.g., fluxes and volumes, respectively) may, according to CF, have to specify the `grid_mapping` again.

But, I fully agree with your proposed approach of storing the CRS with the mesh only:
In my FORTRAN-based `io_netcdf`/`io_ugrid` library I read and store ONE crs structure and associate it 'at top level' with each UGRID mesh, i.e., not store with the nodes or other grid child members.

**Comment by arthurvd** [comment]
> **@bhlevca wrote in the opening message:**
> pyugrid [..] is working great except [..]:
> 1. triangular cells only (but I create triangular meshes only at the moment so I am OK, but I hope for support for other shapes in the near future)triangular cells only (but I create triangular meshes only at the moment so I am OK, but I hope for support for other shapes in the near future)

This is in fact in preparation by @SiggyF in his fork:
https://github.com/pyugrid/pyugrid/compare/master...SiggyF:fedor/flexiblemesh

He can tell better about the current status, and when a PR is appropriate, but don't start from scratch, because something is already in the making. (Maybe any followup discussion deserves a separate issue from the current lat/lon/node issue).

**Comment by bhlevca** [comment]
Arthur is right. I noticed that in `read_netcdf.py` method `load_grid_from_nc_dataset` when loading data is parsing only for `node`  (in var.location) however there are also `face` and `edge` locations for various variables such as velocity.

I believe that most of the required changes are in one method `load_grid_from_nc_dataset`

**Comment by bhlevca** [comment]
@arthurvd Thank you Arthur for pointing that out.  I am not starting anything at the moment in terms of expanding beyond triangles. I will follow what Fedor is doing and perhaps help if necessary.

My first priority is to have working solution for triangular meshes. I hope to pass the parsing phase for data variables. The location attribute is the last hurdle.

**Comment by rsignell-usgs** [comment]
Requiring lat/lon coordinates has always been part of CF.  So even when you have 1D UTM x,y coordinate variables, you need to supply the 2D lat/lon coordinates.   I know these aren't typically stored with Delft3D FM, but could they be?  
The impact on the file size would be tiny, right?

**Comment by ChrisBarker-NOAA** [comment]
"""
Not per se, there's actually also optional edge coordinates and face coordinates, and the data variables defined on those locations (e.g., fluxes and volumes, respectively) may, according to CF, have to specify the grid_mapping again.
"""

right -- I didn't think hard. I'm thinking that we don't need to support having various coordinates with DIFFERENT systems. So there is ONE for the UGrid object, and it can get written to all the coordinate variable when a the file is written. Ideally, when reading, teh file is checked to make sure that they are all compatible.

**Comment by ChrisBarker-NOAA** [comment]
> Requiring lat/lon coordinates has always been part of CF

hmm -- I'm thinking that we may not want to extend this requirement to UGRID -- (and it's note there yet). And it pyugrid can be very useful with other coordinate systems.

But what does pyugrid do if both UTM (for example) AND lat-lon are given?

**Comment by arthurvd** [comment]
> Requiring lat/lon coordinates has always been part of CF. So even with rectilinear grids with 1D UTM x,y coordinate variables, you need to supply the 2D lat/lon coordinate variables. I know these aren't typically stored with Delft3D FM, but could they be? The impact on the file size would be tiny, right?

True, Delft3D-FM still needs to do this. In fact I have it in preparation on my local machine.
Blocker was not file size considerations, but rather: if my input files + model + output files are in a Cartesian/projected coordinate system (e.g. UTM, or some other lesser known location such as Hartebeesthoek94 / Lo23), than having to convert those coordinates to lat/lon during or right after the simulation, solely for compliancy, still feels somewhat awkward.

Fortunately, though, I found out that `fortrangis` currently nicely wraps the PROJ.4 library that can do all this, in fortran interfaces, so I chose to surrender to the conventions and implement coordinate conversion to lat/lon (EPSG:4326).

But again: none of all this should affect UGRID, nor pyugrid.

**Comment by arthurvd** [comment]
> But what does pyugrid do if both UTM (for example) AND lat-lon are given?

I'd say the client application can choose.
When used as input: Delft3D-FM reads an input UGRID file, it will do the following:
* if a grid mapping is found that is NOT EPSG:4326 then it runs the model simulation in Cartesian model (input coordinates in meters)
* otherwise it runs in spherical mode (lat/lon degrees)

When used as output: for postprocessing, both can be fine, matter of preference.
I'll continue part of the lat/lon discussion here:
https://github.com/ugrid-conventions/ugrid-conventions/issues/33#issuecomment-251512515

**Comment by ChrisBarker-NOAA** [comment]
Well, UGrid is the client application. My thought is to say a given UGRid object only has one, and when loading a file, there should be a flag telling it what to load. defaulting to lat-lon.

Alternatively, UGrid could store both, and let the user code work with either, but I'm thinking that's more complication than we'd want.

**Comment by arthurvd** [comment]
Sounds pragmatic: pending any standardization discussions, it is always convenient for the user to optionally enforce a certain coordinate system preference.
